hedera-node/src/main/java/com/hedera/services/store/contracts/precompile/HTSPrecompiledContract.java

```diff
@@ -809,6 +881,245 @@ public Bytes getFailureResultFor(final ResponseCodeEnum status) {
+			 * check if each one specifies the type of key it applies to (that the bit field
+			 * is not `00000000` and no bit bigger than 6 is set) and also that there are not multiple
+			 * keys values for the same key type (e.g. multiple `TokenKey` instances have the adminKey bit set)
+			 */
+			final var tokenKeys = tokenCreateOp.getTokenKeys();
+			if (!tokenKeys.isEmpty()) {
+				for (int i = 0, tokenKeysSize = tokenKeys.size(); i < tokenKeysSize; i++) {
+					final var tokenKey = tokenKeys.get(i);
+					validateTrue(tokenKey.key().getKeyValueType() != INVALID_KEY, INVALID_TRANSACTION_BODY);
+					final var tokenKeyBitField = tokenKey.keyType().intValue();
+					validateTrue(tokenKeyBitField != 0 && tokenKeyBitField < 128, INVALID_TRANSACTION_BODY);
+					for (int j = i + 1; j < tokenKeysSize; j++) {
+						validateTrue((tokenKeyBitField & tokenKeys.get(j).keyType().intValue()) == 0,
+								INVALID_TRANSACTION_BODY);
+					}
+				}
+			}
+
+			/*
+			 * The denomination of a fixed fee depends on the values of tokenId, useHbarsForPayment
+			 * useCurrentTokenForPayment. Exactly one of the values of the struct should be set.
+			 */
+			if (!tokenCreateOp.getFixedFees().isEmpty()) {
+				for (final var fixedFee: tokenCreateOp.getFixedFees()) {
+					validateTrue(fixedFee.getFixedFeePayment() != INVALID_PAYMENT, INVALID_TRANSACTION_BODY);
+				}
+			}
+
+			/*
+			 * When a royalty fee with fallback fee is specified, we need to check that
+			 * the fallback fixed fee is valid.
+			 */
+			if (!tokenCreateOp.getRoyaltyFees().isEmpty()) {
+				for (final var royaltyFee: tokenCreateOp.getRoyaltyFees()) {
+					validateTrue(royaltyFee.fallbackFixedFee().getFixedFeePayment() != INVALID_PAYMENT, INVALID_TRANSACTION_BODY);
+				}
+			}
+		}
+
+		private void replaceInheritedKeysWithSenderKey() throws DecoderException {
+			tokenCreateOp.setAllInheritedKeysTo((JKey) ledgers.accounts().get(
+					EntityIdUtils.accountIdFromEvmAddress(senderAddress), AccountProperty.KEY));
+		}
+
+		private boolean validateAdminKey(
+				final MessageFrame frame,
+				final TokenCreateWrapper.TokenKeyWrapper tokenKeyWrapper
+		) {
+			final var key = tokenKeyWrapper.key();
+			return switch (key.getKeyValueType()) {
+				case INHERIT_ACCOUNT_KEY -> validateKey(frame, senderAddress, sigsVerifier::hasActiveKey);
+				case CONTRACT_ID -> validateKey(frame, asTypedEvmAddress(key.getContractID()),
+								sigsVerifier::hasActiveKey);
+				case DELEGATABLE_CONTRACT_ID -> validateKey(frame, asTypedEvmAddress(key.getDelegatableContractID()),
+								sigsVerifier::hasActiveKey);
+				case ED25519 -> validateCryptoKey(new JEd25519Key(key.getEd25519Key()),
+								sigsVerifier::cryptoKeyIsActive);
+				case ECDS_SECPK256K1 -> validateCryptoKey(new JECDSASecp256k1Key(key.getEcdsSecp256k1()),
+								sigsVerifier::cryptoKeyIsActive);
+				default -> throw new InvalidTransactionException(FAIL_INVALID);
```

Comment:
As Stoyan noted above, default error codes would be e.g. `NOT_SUPPORTED` instead of `FAIL_INVALID` (which should mean a bug in our code).